### ISSUE TITLE
Too many threads

### ISSUE THREAD

**Issue by D-AIRY** [SmartlyDressedGames/Unturned-3.x-Community#524]
Empty vanilla server with PEI map spawns 165 threads on startup. Is it expected behaviour?

**Comment by Trojaner** [comment]
The amount of threads does not say anything about software at all.
For example, my computer currently has 3522 active threads distributed from 267 processes.

Unity 5 usually only has one active game logic thread and Unturned (afaik) uses another thread for networking. All other threads are likely internal threads created and used by the Unity Engine. These are very likely not active all the time and/or are waiting for events (e.g. input, graphics, WinAPI stuff).

tl;dr: Most of these threads are likely not active and instead only wait for certain things to happen.
This is pretty normal.

**Comment by D-AIRY** [comment]
I don't think so. There high cpu usage and seems almost half of time is spent in synchronization.
There are 9173 threads in 41 unturned servers. It much more than ~13 threads per process from your example

**Comment by D-AIRY** [comment]
Count of spawned thread depends on visible number of cpu cores for process. If run the same server with 1 visible core, it runs only 57 threads.

**Comment by Trojaner** [comment]
> There are 9173 threads in 41 unturned servers. It much more than ~13 threads per process from your example
Yes but your average process is not a game which requires way more than your average process resources.

> I don't think so. There high cpu usage and seems almost half of time is spent in synchronization.
High CPU usage is sometimes a problem on Unturned, yes, **but it is definitely not bound to amount of created threads.** It's related to ***active*** threads and what exactly these threads are doing. Like said, most of these threads are not actively doing something. If there is an synchronization issue like you mentioned; I assume it's just because of the huge amount of servers you are apparently running. What are your server specs?

The amount of threads created are also more related to the engine, Unturned itself does almost not utilize multihreading at all (it can't anyway in Unity 5).

**Comment by D-AIRY** [comment]
2x Intel xeon e5 2690 v3 with 64GB RAM, running ubuntu 14.04
Avg cpu usage ~70%, half of it - kernel time. LA is about 100, disk subsystem is not almost non loaded, so it can't cause such LA.
strace of one of the server shows that there a lot of time spent on futex() syscall, which generally should run in userspace

**Comment by Trojaner** [comment]
You are right about context switches being the issue, but like said this seems to be related to the huge amount of Unturned servers you are running on a single machine. 24 + 24 threads are simply not enough to run 41 Unturned servers (even at only 2 active threads per game process this would cause a lot of context switches on your machine), I highly doubt this issue can be addressed by Unturned or Unity.

**Comment by SDGNelson** [comment]
I'm inclined to agree with Trojaner that it's an expected number of threads. Game-side Unturned has its game logic thread which is typical, and some minor multi-threading for a couple things. The engine itself of course has multiple threads. Using the steam SDK introduces more threads, and as part of that more threads for sockets / networking as Trojaner mentioned.
